gnd/src/main/java/com/google/android/gnd/ui/editobservation/PhotoFieldViewModel.java

```diff
@@ -0,0 +1,80 @@
+/*
+ * Copyright 2020 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.android.gnd.ui.editobservation;
+
+import android.net.Uri;
+import android.view.View;
+import androidx.databinding.ObservableMap;
+import androidx.databinding.ObservableMap.OnMapChangedCallback;
+import androidx.lifecycle.LiveData;
+import androidx.lifecycle.MutableLiveData;
+import com.google.android.gnd.model.form.Field;
+import com.google.android.gnd.model.observation.Response;
+import com.google.android.gnd.system.StorageManager;
+import com.google.android.gnd.ui.common.AbstractViewModel;
+import javax.inject.Inject;
+
+public class PhotoFieldViewModel extends AbstractViewModel {
+
+  private final StorageManager storageManager;
+  private final MutableLiveData<Uri> destinationPath = new MutableLiveData<>();
+  private final MutableLiveData<Integer> photoPreviewVisibility = new MutableLiveData<>(View.GONE);
+
+  @Inject
+  PhotoFieldViewModel(StorageManager storageManager) {
+    this.storageManager = storageManager;
+  }
+
+  public LiveData<Uri> getDestinationPath() {
+    return destinationPath;
+  }
+
+  public MutableLiveData<Integer> photoPreviewVisibility() {
+    return photoPreviewVisibility;
+  }
+
+  public void init(Field field, ObservableMap<String, Response> responses) {
+    // Load last saved value
+    updateUI(responses.get(field.getId()), field);
+
+    // Observe response updates
+    responses.addOnMapChangedCallback(
+        new OnMapChangedCallback<ObservableMap<String, Response>, String, Response>() {
+          @Override
+          public void onMapChanged(ObservableMap<String, Response> sender, String key) {
+            if (key.equals(field.getId())) {
+              updateUI(sender.get(key), field);
+            }
+          }
+        });
+  }
+
+  private void updateUI(Response response, Field field) {
+    if (response == null) {
+      photoPreviewVisibility.setValue(View.GONE);
+    } else {
+      String value = response.getDetailsText(field);
+      if (value.isEmpty()) {
+        photoPreviewVisibility.setValue(View.GONE);
+      } else {
+        photoPreviewVisibility.setValue(View.VISIBLE);
+        disposeOnClear(
```

Comment:
This will accumulate subscriptions in the viewmodel ad infinitum every time a photo is changed during the viewmodel's lifecycle. That's probably ok in practice since the user will never likely do this so many times we'll run out of memory, but it is a form of memory leak that as a general rule we're designing to avoid.

The "right", safe way to do this would be to create a `PublishSubject<Optional<String>> destinationPath` class member that emits the path or empty when this method is called. Then in the constructor, you would define `LiveData<Optional<Uri>> uri` using something like:
```java
  this.uri = LiveDataReactiveStreams.fromPublisher(
    destinationPath.switchMap(p -> p.map(storageManager::loadUriFromDestinationPath)));
```

Visibility could be defined in a similar way:
```java
  this.uri = LiveDataReactiveStreams.fromPublisher(
    destinationPath.map(p -> p.isPresent() ? View..VISIBLE : View.GONE));
```

This method would then simply call `destinationPath.onNext` with either `Optional.empty()` or the URI.

Wdyt?

Reply:
Looks so clean. ✨

Reply:
🌟

Reply:
I am having a hard time creating the `uri` LiveData object.

For visibility, the below code works
```
    photoPreviewVisibility =
        LiveDataReactiveStreams.fromPublisher(
            destinationPath
                .map(path -> path.isPresent() ? View.VISIBLE : View.GONE)
                .toFlowable(BackpressureStrategy.LATEST));
```

But for Uri, I can't get the correct RxChain

`loadUriFromDestinationPath(path)` returns `Single<Uri>`

Reply:
If you replace destinationPath with a `PublishProcessor` you can get rid of the .toFlowable. Looking into the right syntax for your case now.

Reply:
To be precise, I can't convert Single<Uri> to Optional<Uri> in the below chain. The only way I can think of is by subscribing `loadUriFromDestinationPath` inside `map` but that would defeat the entire purpose of this refactor.

```
    uri =
        LiveDataReactiveStreams.fromPublisher(
            Flowable.fromPublisher(destinationPath)
                .map(Optional::get)
                .map(storageManager::loadUriFromDestinationPath));
```